### PR TITLE
Fix #1023

### LINKED ISSUE #1023: Taskqueue can segfault
As shown here https://github.com/STORM-IRIT/Radium-Engine/actions/runs/3463313657/jobs/5783393792
Taskqueue unittest randomly segfault.

### PR DESCRIPTION
**Pull request by dlyr**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

Be aware that the PR request cannot be accepted if it doesn't pass the Continuous Integration tests.

- **What kind of change does this PR introduce?**
  - [x] bug fix
  - [ ] feature
  - [ ] docs update
  - [ ] other:

- **What is the current behavior?** (You can also link to an open issue here)
There is a concurrency problem in task queue, when worker idle wainting for task, they can steal task from runTaskInThisThread.
Moreover, if registerTask is not renetrant, and can cause task incoherency if task are added to one queue from different threads.
Fix #1023

### CODE REVIEW

File: src/Core/Tasks/TaskQueue.cpp [hunk 1]
```diff
@@ -56,6 +58,7 @@ void TaskQueue::addDependency( TaskQueue::TaskId predecessor, TaskQueue::TaskId
                             successor ) == m_dependencies[predecessor].end(),
                  "Cannot add a dependency twice" );
 
+    std::lock_guard<std::mutex> lock( m_taskMutex );
```
Comment (nmellado) [inline src/Core/Tasks/TaskQueue.cpp:61]: Maybe move before asserts ?

File: src/Core/Tasks/TaskQueue.hpp [hunk 2]
```diff
@@ -43,7 +43,8 @@ class RA_CORE_API TaskQueue
     };
 
   public:
-    /// Constructor. Initializes the thread pools with numThreads threads.
+    /// Constructor. Initializes the thread worker pools with numThreads threads.
+    /// if numThreads == 0, its a runTasksInThisThread only task queu
```
Comment (nmellado) [inline src/Core/Tasks/TaskQueue.hpp:47]:
```suggestion
    /// if numThreads == 0, its a runTasksInThisThread only task queue
```

**Review by nmellado — COMMENTED** [review]
Sounds good overall (see minor comments)

File: src/Core/Tasks/TaskQueue.hpp [hunk 3]
```diff
@@ -146,6 +147,9 @@ class RA_CORE_API TaskQueue
     std::condition_variable m_threadNotifier;
     /// Global mutex over thread-sensitive variables.
     std::mutex m_taskQueueMutex;
+    /// Mutex for task registration (m_tasks, m_dependencies, m_timerData ...) if task queue is
+    /// fed from multiples threads
```
[the remark this answers is not in the document]
Comment (dlyr) [reply src/Core/Tasks/TaskQueue.hpp:151]: to feed a queue ?

[the remark this answers is not in the document]
Comment (dlyr) [reply src/Core/Tasks/TaskQueue.hpp:151]: if task are registered from multiple threads, ?

[the remark this answers is not in the document]
Comment (nmellado) [reply src/Core/Tasks/TaskQueue.hpp:151]: Bon pour moi:
If tasks are registered from multiple threads

**Review by hiergaut — APPROVED** [review]
Must be tested on all supported platforms.

**Comment by hiergaut** [comment]
The CI jobs was cancelled on macOS, how to get these jobs going again ?

**Comment by dlyr** [comment]
It seems there is a problem with unittests on macos, tests last forever, that's why they are cancelled

**Comment by dlyr** [comment]
It seems there is a problem with unittests on macos, tests last forever, that's why they are cancelled

**Review by MathiasPaulin — APPROVED** [review]

**Pull request merged**